keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,1132 @@
+  capabilities that are only available in the initial (host) user namespace such
+  as `CAP_SYS_TIME`, `CAP_SYS_MODULE` & `CAP_MKNOD`.
+
+  If a pod is given one of those capabilities it will still be deployed, but the
+  capability will be ineffective and processes using those capabilities will
+  fail. This is not impacting the implementation in Kubernetes. If users need
+  the capability to be effective, they should use `userNamespaceMode=Host`.
+
+  The list of such capabilities is likely to change from one Linux version to
+  another. For example, Linux now has [time
+  namespaces](https://man7.org/linux/man-pages/man7/time_namespaces.7.html) and
+  there are ways to make `CAP_SYS_TIME` work inside a user namespace. There are
+  also discussions to make `CAP_MKNOD` work in user namespaces.
+
+- **Sharing Host Namespaces**:
+  There are some limitations in the Linux kernel and in the runtimes that
+  prevent sharing other host namespaces when the host user namespace is not
+  shared.
+  - Mounting `mqueue` (`/dev/mqueue`) is not allowed from a process in a user
+    namespace that does not own the IPC namespace. Pods with `hostIPC=true` and
+    `userNamespaceMode=Pod|Cluster` can fail.
+  - Mounting `procfs` (`/proc`) is not allowed from a process in a user
+    namespace that does not own the PID namespace. Pods with `hostPID=true` and
+    `userNamespaceMode=Pod|Cluster` can fail.
+  - Mounting `sysfs` (`/sys`) is not allowed from a process in a user namespace
+    that does not own the network namespace. Impact: pods with
+    `hostNetwork=true` and `userNamespaceMode=Pod|Cluster` can fail.
+
+  If users specify `userNamespaceMode=Pod|Cluster` and one of these
+  `host{IPC,PID,Network}=true` options, runc will currently fail to start the
+  container. The kubelet does **not** try to prevent that combination of
+  options, in case runc or the kernel make it possible in the future to use that
+  combination.
+
+In order to avoid breaking existing workloads `Host` is the default value of
+`userNamespaceMode`.
+
+#### Duplicated Snapshots of Container Images
+
+The owners of the files of a container image have to been set accordingly to the
+ID mappings used for that container. For example, if the user 0 in the container
+is mapped to the host user 100000, then the `/root` directory has to be owned by
+user ID 100000 in the host, so it appears to belong to root in the container.
+The current implementation in container runtimes is to recursively perform a
+`chown` operation over the image snapshot when it's pulled. This presents a risk
+as it potentially increases the time and the storage needed to handle the
+container images.
+
+Some runtimes, like cri-o, mitigate these problems by using the `metacopy`
+option of overlayfs. This option avoids copying the whole file content when an
+operation updating only the metadata, like `chown` or `chmod`, is performed.
+This solution could be adopted by other runtimes until a more sophisticated
```

Comment:
```
Metadata only copy up
--------------------

When metadata only copy up feature is enabled, overlayfs will only copy
up metadata (as opposed to whole file), when a metadata specific operation
like chown/chmod is performed. Full file will be copied up later when
file is opened for WRITE operation.
```
Container/storage used by CRI-O mounts an overlay file system with the metacopy=on flag set, it then chowns all of the lower files in the image to match the user namespace to which the container will run. This operation is very quick compared to standard chowning, since none of the files data has to be copied up.  If a second container runs on the same image with the same user namespace, then the chowned image is shared, eliminating the need to chown again.